### ISSUE TITLE
Training on a custom dataset (Fine tuning)

### ISSUE THREAD

**Issue by MyVanitar** [DavideA/deeplabv2-keras#1]
Hello,

1) have you tested the IOU or accuracy of your repo on VOC?

2) How can I train (fine-tune) the network on the custom dataset with different number of classes?

**Comment by DavideA** [comment]
Hey.

1. Yes I did. The model performed as reported in the paper on VOC12 segmentation benchmark.

2. In my opinion, the most straightforward way to fine tune the model with a different number of classes is to change the last convolution of each layer of the atrous spatial pyramid (namely layers `fc8_voc12_1`, `fc8_voc12_2`, `fc8_voc12_3`, `fc8_voc12_4`) and set the number of output channels as the number of classes you need. Doing that will of course cause an exception in loading weights (line 211 of `deeplabv2.py`). You can change the name of new layers and set `by_name` parameter to `True` in `load_weights`. This way, weights will be loaded for all layers but the ones you modified. You can now either train the last convolution only (setting all other layers as non trainable) or let the gradient flow in the whole network. 

Hope this helps!
Let me know how it goes!

D

**Comment by MyVanitar** [comment]
Thanks for the reply.

For fine-tuning we should have the pre-trained model as well which here should be in Keras format which I don't have information about it.

if I understand the process it should be like this:

1) change the number of classes in these layers: `fc8_voc12_1, fc8_voc12_2, fc8_voc12_3, fc8_voc12_4`
2) renaming the new layers and also set the `by_name` value to `True`.

from here I didn't understand. first which layers are new layers and second what is the fine-tuning process afterwards.

I really appreciate if you make a test of this yourself because this is a very important topic and beneficial for all. Your repo in Keras is unique here.

**Comment by DavideA** [comment]
"new layers" are the ones you modified to match the desired number of classes. If you rename such layers and set `by_name` to `True` you will still have the deeplab model, but the layers you modified are new born and randomly initialized.

You can now fine-tune the model with new data using the Keras training routines. 

D

**Comment by MyVanitar** [comment]
1) where should I include the pretrained weights? I think the original weights are in `.caffemodel` format

2) is there any script for this fine-tuning purpose in Keras?

**Comment by DavideA** [comment]
1. I converted all the original weights from the `.caffemodel` into the Keras format (Theano backend only). You can download such weight file [here](http://imagelab.ing.unimore.it/files/deeplabV2_weights/deeplabV2_weights_th.h5), but there is no need to do it, my code will take care of it.

2. Fine tuning is basically a synonym for "taking a network and training it on new data". You can search in the [Keras model docs](https://keras.io/models/model/) how `fit`, `fit_generator` and `train_on_batch` functions work. You could also be interested in the [Keras examples](https://github.com/fchollet/keras/tree/master/examples).

D

**Comment by MyVanitar** [comment]
Should I make these modifications in `deeplabV2.py`?

**Comment by DavideA** [comment]
Yes, the layers modification should be made in `deeplabV2.py`. I would code the fine tuning script as a separate file.

**Comment by MyVanitar** [comment]
Have you made that separate file for fine-tuning?

**Comment by DavideA** [comment]
Nope. I didn't and I won't.

Fine tuning equals training and how to train a keras model is deeply explained in the documentation.

Closing.
D

**Comment by MyVanitar** [comment]
Okay, take it easy. I asked because you had promised :-)
